File: web/js/render.js
// Annotated-frame overlay, drawn with the 2D canvas API (mirrors escalator_monitor/render.py).
import { State, quadRegions } from "./core.js";

// Status colours (fixed across themes): good / critical / warning / muted.
export const STATE_COLORS = {
  [State.WORKING]: "#0ca30c",
  [State.STOPPED]: "#d03b3b",
  [State.IDLE]: "#fab219",
  [State.INITIALIZING]: "#898781",
};
const STATE_TEXT = {
  [State.WORKING]: "#ffffff",
  [State.STOPPED]: "#ffffff",
  [State.IDLE]: "#1a1a19",
  [State.INITIALIZING]: "#ffffff",
};
const MOVING_TINT = "rgba(12, 163, 12, 0.18)";
const STILL_TINT = "rgba(236, 131, 90, 0.18)";
const PERSON = "#2ee66b";
const FONT = 'system-ui, -apple-system, "Segoe UI", sans-serif';

function path(ctx, poly) {
  ctx.beginPath();
  poly.forEach(([x, y], i) => (i ? ctx.lineTo(x, y) : ctx.moveTo(x, y)));
  ctx.closePath();
}

function pill(ctx, x, y, w, h, r) {
  ctx.beginPath();
  ctx.roundRect(x, y, w, h, r);
}

export class Renderer {
  constructor(cfg, quad, width, height) {
    this.cfg = cfg;
    this.width = width;
    this.height = height;
    this.ui = Math.min(2.5, Math.max(0.7, Math.min(width, height) / 540));
    this.setRoi(quad);
  }

  setRoi(quad) {
    this.quad = quad;
    this.regions = quadRegions(quad, this.cfg.handrailWidthFrac);
  }

  /** Draw `frame` (a canvas) plus the overlay for result `r` into ctx. */
  draw(ctx, frame, r, duration = 0) {
    const { width: w, height: h, ui } = this;
    ctx.drawImage(frame, 0, 0, w, h);
    const color = STATE_COLORS[r.state];

    if (r.motion.valid) {
      ctx.fillStyle = r.motion.handrailScore > 0.3 ? MOVING_TINT : STILL_TINT;
      path(ctx, this.regions.left);
      ctx.fill();
      path(ctx, this.regions.right);
      ctx.fill();
      ctx.fillStyle = r.motion.stepsScore > 0.3 ? MOVING_TINT : STILL_TINT;
      path(ctx, this.regions.steps);
      ctx.fill();
    }

    ctx.lineWidth = Math.max(1.5, 2 * ui);
    ctx.strokeStyle = PERSON;
    ctx.font = `600 ${Math.round(11 * ui)}px ${FONT}`;
    ctx.fillStyle = PERSON;
    for (const p of r.people) {
      const [x1, y1, x2, y2] = p.box;
      ctx.strokeRect(x1, y1, x2 - x1, y2 - y1);
      if (p.conf < 1) ctx.fillText(p.conf.toFixed(2), x1, Math.max(10, y1 - 4));
    }

    ctx.lineWidth = Math.max(2, 2.5 * ui);
    ctx.strokeStyle = color;
    path(ctx, this.quad);
    ctx.stroke();

    this.badge(ctx, r);
    if (this.cfg.showDebug) this.panel(ctx, r);
    if (duration > 0) {
      ctx.fillStyle = color;
      ctx.fillRect(0, h - Math.max(3, 4 * ui), (w * Math.min(1, r.time / duration)) | 0, Math.max(3, 4 * ui));
    }
  }

  badge(ctx, r) {
    const { width: w, ui } = this;
    const labels = [[r.state, STATE_COLORS[r.state], STATE_TEXT[r.state]]];
    if (r.wrongDirection) labels.push(["WRONG DIRECTION", STATE_COLORS[State.STOPPED], "#ffffff"]);
    ctx.font = `700 ${Math.round(17 * ui)}px ${FONT}`;
    ctx.textBaseline = "middle";
    let y = 12 * ui;
    for (const [text, bg, fg] of labels) {
      const tw = ctx.measureText(text).width;
      const bw = tw + 28 * ui;
      const bh = 34 * ui;
      const x = (w - bw) / 2;
      ctx.globalAlpha = 0.94;
      ctx.fillStyle = bg;
      pill(ctx, x, y, bw, bh, 8 * ui);
      ctx.fill();
      ctx.globalAlpha = 1;
      ctx.fillStyle = fg;
      ctx.fillText(text, x + 14 * ui, y + bh / 2 + 1);
      y += bh + 6 * ui;
    }
    ctx.textBaseline = "alphabetic";
  }

  panel(ctx, r) {
    const { height: h, ui } = this;
    const m = r.motion;
    const minutes = Math.floor(r.time / 60);
    const seconds = (r.time % 60).toFixed(2).padStart(5, "0");
    const lines = [
      [`Time ${String(minutes).padStart(2, "0")}:${seconds}   People in ROI ${r.people.length}`, "#ffffff"],
      [`Handrails  mag ${m.handrailMag.toFixed(2)}  cons ${m.handrailCons.toFixed(2)}  score ${m.handrailScore.toFixed(2)}`, m.handrailScore > 0.3 ? "#a8f0a8" : "#c9c9c9"],
      [`Steps      mag ${m.stepsMag.toFixed(2)}  cons ${m.stepsCons.toFixed(2)}  score ${m.stepsScore.toFixed(2)}`, m.stepsScore > 0.3 ? "#a8f0a8" : "#c9c9c9"],
      [`Move confidence ${m.confidence.toFixed(2)} (need ${this.cfg.moveConfidenceMin.toFixed(2)})  moving ${Math.round(100 * r.window.movingRatio)}%`, "#ffffff"],
    ];
    if (r.direction) lines.push([`Surface moving ${r.direction} (image)`, "#c9c9c9"]);
    ctx.font = `500 ${Math.round(11.5 * ui)}px ui-monospace, SFMono-Regular, Menlo, monospace`;
    const lineH = 17 * ui;
    const width = Math.max(...lines.map(([t]) => ctx.measureText(t).width)) + 20 * ui;
    const height = lines.length * lineH + 12 * ui;
    const x = 10 * ui;
    const y = h - height - 14 * ui;
    ctx.fillStyle = "rgba(15, 15, 15, 0.78)";
    pill(ctx, x, y, width, height, 6 * ui);
    ctx.fill();
    lines.forEach(([text, c], i) => {
      ctx.fillStyle = c;
      ctx.fillText(text, x + 10 * ui, y + 6 * ui + lineH * (i + 0.8));
    });
  }
}

File: escalator_monitor/render.py
"""Overlay drawing for the annotated output video."""

from __future__ import annotations

import cv2
import numpy as np

from .config import Config
from .geometry import Quad, rasterize
from .pipeline import FrameResult
from .state import State

FONT = cv2.FONT_HERSHEY_SIMPLEX

# BGR fill colour and text colour per state.
STATE_STYLE: dict[State, tuple[tuple[int, int, int], tuple[int, int, int]]] = {
    State.WORKING: ((0, 160, 0), (255, 255, 255)),
    State.STOPPED: ((0, 0, 205), (255, 255, 255)),
    State.IDLE: ((0, 190, 255), (20, 20, 20)),
    State.INITIALIZING: ((120, 120, 120), (255, 255, 255)),
}
MOVING_TINT = (0, 220, 0)
STILL_TINT = (0, 100, 220)
PERSON_COLOR = (0, 220, 0)


def blend_rect(img: np.ndarray, x1: int, y1: int, x2: int, y2: int, color, alpha: float) -> None:
    """Blend a solid rectangle into ``img`` in place, clipped to the image."""
    h, w = img.shape[:2]
    x1, y1, x2, y2 = max(0, x1), max(0, y1), min(w, x2), min(h, y2)
    if x2 <= x1 or y2 <= y1:
        return
    roi = img[y1:y2, x1:x2]
    cv2.addWeighted(np.full_like(roi, color), alpha, roi, 1 - alpha, 0, dst=roi)


class Renderer:
    def __init__(self, cfg: Config, quad: Quad, frame_size: tuple[int, int]):
        self.cfg = cfg
        self.w, self.h = frame_size
        self.ui = float(np.clip(min(self.w, self.h) / 720, 0.5, 2.5))
        self.thick = max(1, round(1.6 * self.ui))
        self.set_roi(quad)

    def set_roi(self, quad: Quad) -> None:
        self.quad = quad
        x0, y0, x1, y1 = quad.bbox(0.0, self.w, self.h)
        self.box = (x0, y0, x1, y1)
        shape = (y1 - y0, x1 - x0)
        polys = quad.regions(self.cfg.handrail_width_frac)
        rails = rasterize(polys["left"], shape, (x0, y0)) | rasterize(polys["right"], shape, (x0, y0))
        self.rail_mask = rails.astype(bool)
        self.steps_mask = rasterize(polys["steps"], shape, (x0, y0)).astype(bool)
        self.region_mask = (rails | self.steps_mask).astype(np.uint8)
        self._layers: dict[tuple[bool, bool], np.ndarray] = {}

    # ------------------------------------------------------------------------
    def draw(self, frame: np.ndarray, r: FrameResult, total_frames: int = 0, proc_fps: float = 0.0) -> np.ndarray:
        out = frame.copy()
        fill, _ = STATE_STYLE[r.state]
        if r.motion.valid:
            self._tint_regions(out, r)
        self._draw_people(out, r)
        cv2.polylines(out, [self.quad.as_int()], True, fill, self.thick + 1, cv2.LINE_AA)
        tl = self.quad.as_int()[0]
        cv2.putText(
            out,
            "Escalator ROI",
            (int(tl[0]), max(12, int(tl[1] - 8 * self.ui))),
            FONT,
            0.55 * self.ui,
            fill,
            self.thick,
            cv2.LINE_AA,
        )
        self._draw_badge(out, r)
        if self.cfg.show_debug_overlay:
            self._draw_panel(out, r, total_frames, proc_fps)
        if total_frames > 0:
            bar = int(self.w * min(1.0, r.index / total_frames))
            cv2.rectangle(out, (0, self.h - max(3, int(4 * self.ui))), (bar, self.h), fill, -1)
        return out

    def _tint_layer(self, rails_moving: bool, steps_moving: bool) -> np.ndarray:
        key = (rails_moving, steps_moving)
        if key not in self._layers:
            layer = np.zeros((*self.rail_mask.shape, 3), np.uint8)
            layer[self.rail_mask] = MOVING_TINT if rails_moving else STILL_TINT
            layer[self.steps_mask] = MOVING_TINT if steps_moving else STILL_TINT
            self._layers[key] = layer
        return self._layers[key]

    def _tint_regions(self, out: np.ndarray, r: FrameResult) -> None:
        x0, y0, x1, y1 = self.box
        roi = out[y0:y1, x0:x1]
        layer = self._tint_layer(r.motion.handrail_score > 0.3, r.motion.steps_score > 0.3)
        blended = cv2.addWeighted(layer, 0.14, roi, 0.86, 0)
        roi[...] = cv2.copyTo(blended, self.region_mask, roi.copy())  # blend inside the ROI only

    def _draw_people(self, out: np.ndarray, r: FrameResult) -> None:
        for track in r.people:
            d = track.det
            cv2.rectangle(out, (d.x1, d.y1), (d.x2, d.y2), PERSON_COLOR, self.thick)
            if d.conf < 1.0:
                cv2.putText(
                    out,
                    f"{d.conf:.2f}",
                    (d.x1, max(10, d.y1 - 4)),
                    FONT,
                    0.45 * self.ui,
                    PERSON_COLOR,
                    self.thick,
                    cv2.LINE_AA,
                )

    def _draw_badge(self, out: np.ndarray, r: FrameResult) -> None:
        labels = [(r.state.value, *STATE_STYLE[r.state])]
        if r.wrong_direction:
            labels.append(("WRONG DIRECTION", (0, 0, 205), (255, 255, 255)))
        y = int(15 * self.ui)
        scale = 0.85 * self.ui
        pad = int(14 * self.ui)
        for text, fill, fg in labels:
            (tw, th), _ = cv2.getTextSize(text, FONT, scale, self.thick + 1)
            x1 = (self.w - tw) // 2 - pad
            x2, y2 = x1 + tw + 2 * pad, y + th + 2 * pad
            blend_rect(out, x1, y, x2, y2, fill, 0.88)
            cv2.rectangle(out, (max(0, x1), y), (min(self.w - 1, x2), y2), (255, 255, 255), max(1, self.thick - 1))
            cv2.putText(out, text, (x1 + pad, y2 - pad), FONT, scale, fg, self.thick + 1, cv2.LINE_AA)
            y = y2 + int(8 * self.ui)

    def _draw_panel(self, out: np.ndarray, r: FrameResult, total_frames: int, proc_fps: float) -> None:
        m, cfg = r.motion, self.cfg
        white, dim, good, blue = (255, 255, 255), (200, 200, 200), (180, 255, 180), (255, 210, 200)
        minutes, seconds = divmod(r.time, 60)
        frame_txt = f"Frame {r.index}/{total_frames}" if total_frames > 0 else f"Frame {r.index}"
        lines = [
            (f"{frame_txt}  Time {int(minutes):02d}:{seconds:05.2f}", white),
            (f"People in ROI: {len(r.people)}  (window {r.window.people_ratio:.0%})", white),
            (
                f"Handrail mag={m.handrail_mag:.2f} cons={m.handrail_cons:.2f} score={m.handrail_score:.2f}",
                good if m.handrail_score > 0.3 else dim,
            ),
            (
                f"Steps    mag={m.steps_mag:.2f} cons={m.steps_cons:.2f} score={m.steps_score:.2f}",
                good if m.steps_score > 0.3 else dim,
            ),
            (f"Direction gate {m.direction_gate:.2f}  rail agreement {m.handrail_agreement:.2f}", blue),
            (
                f"Move confidence {m.confidence:.2f} (need {cfg.move_confidence_min:.2f})  moving {r.window.moving_ratio:.0%}",
                white,
            ),
        ]
        extra = []
        if r.direction:
            extra.append(f"Surface moving {r.direction} (image)")
        if cfg.compensate_camera_motion:
            extra.append(f"camera shift {m.camera_shift[0]:+.2f},{m.camera_shift[1]:+.2f}")
        if proc_fps > 0:
            extra.append(f"{proc_fps:.1f} FPS")
        if extra:
            lines.append(("  ".join(extra), (170, 170, 170)))

        scale = 0.5 * self.ui
        line_h = int(24 * self.ui)
        width = max(cv2.getTextSize(t, FONT, scale, self.thick)[0][0] for t, _ in lines) + int(28 * self.ui)
        height = line_h * len(lines) + int(14 * self.ui)
        x1 = int(12 * self.ui)
        y1 = max(0, self.h - height - int(16 * self.ui))
        blend_rect(out, x1, y1, x1 + width, y1 + height, (15, 15, 15), 0.8)
        for i, (text, color) in enumerate(lines):
            cv2.putText(
                out, text, (x1 + int(14 * self.ui), y1 + line_h * (i + 1)), FONT, scale, color, self.thick, cv2.LINE_AA
            )
